integrations/spark/spark-3.5/openhouse-spark-itest/src/test/java/com/linkedin/openhouse/spark/catalogtest/BranchTestSpark3_5.java

```diff
@@ -0,0 +1,2367 @@
+
+      List<Row> branchAData =
+          spark
+              .sql("SELECT name FROM " + tableName + " VERSION AS OF 'branchA' ORDER BY name")
+              .collectAsList();
+      assertEquals(
+          "branchA_data", branchAData.get(0).getString(0), "BranchA should contain branchA_data");
+      assertEquals(
+          "main_data", branchAData.get(1).getString(0), "BranchA should contain main_data");
+
+      List<Row> branchBData =
+          spark
+              .sql("SELECT name FROM " + tableName + " VERSION AS OF 'branchB' ORDER BY name")
+              .collectAsList();
+      assertEquals(
+          "branchA_data", branchBData.get(0).getString(0), "BranchB should contain branchA_data");
+      assertEquals(
+          "branchB_data", branchBData.get(1).getString(0), "BranchB should contain branchB_data");
+      assertEquals(
+          "main_data", branchBData.get(2).getString(0), "BranchB should contain main_data");
+
+      // ===== TEST THE SPECIFIC SCENARIO THAT WOULD HAVE BEEN AMBIGUOUS =====
+
+      // At this point, we have:
+      // - main points to snapshotX
+      // - branchA points to snapshotY
+      // - branchB points to snapshotZ
+      //
+      // If we were to commit a new snapshot as child of snapshotY, our fixed logic should work
+      // because only the explicitly targeted branch (via branch-specific insert syntax) should be
+      // considered
+
+      // Verify that we can still commit to branchA even though multiple branches exist
+      spark.sql("INSERT INTO " + tableName + ".branch_branchA VALUES ('additional_branchA_data')");
+
+      // Verify branchA advanced but branchB didn't
+      List<Row> branchAFinalSnapshots2 =
+          spark
+              .sql("SELECT snapshot_id FROM " + tableName + ".refs WHERE name = 'branchA'")
+              .collectAsList();
+      List<Row> branchBFinalSnapshots2 =
+          spark
+              .sql("SELECT snapshot_id FROM " + tableName + ".refs WHERE name = 'branchB'")
+              .collectAsList();
+
+      long finalSnapshotY2 = branchAFinalSnapshots2.get(0).getLong(0);
+      long finalSnapshotZ2 = branchBFinalSnapshots2.get(0).getLong(0);
+
+      assertNotEquals(snapshotY, finalSnapshotY2, "BranchA should have advanced to a new snapshot");
+      assertEquals(snapshotZ, finalSnapshotZ2, "BranchB should remain at the same snapshot");
+
+      // Verify data counts
+      assertEquals(
+          3,
+          spark
+              .sql("SELECT * FROM " + tableName + " VERSION AS OF 'branchA'")
+              .collectAsList()
+              .size(),
+          "BranchA should now have 3 rows");
+      assertEquals(
+          3,
+          spark
+              .sql("SELECT * FROM " + tableName + " VERSION AS OF 'branchB'")
+              .collectAsList()
+              .size(),
+          "BranchB should still have 3 rows (unchanged)");
+    }
+  }
+
+  // ===== CHERRY PICKING BETWEEN BRANCHES =====
```

Comment:
Checking my understanding. So wap snapshots only base off main branch (their parents will be main branch even though we are in feature branch now), and cherrypick only applies to main branch. Wap.id and branching seems to be orthogonal concepts?